### ISSUE TITLE
Atomic metadata update api lookup fails

### ISSUE THREAD

**Issue by mksanger** [irods/irods#6004]
### iRODS Version, OS and Version
iRODS 4.2.10, ubuntu 18.04

### What did you try to do?
Perform an atomic metadata update using the C api in the following program:
```
#include <rodsClient.h>

#include <atomic_apply_metadata_operations.h> //not available through apiHeaderAll.h

int main() {
    char *json_str = 
        "{\"entity_name\": \"/testZone/home/irods/test_object\", \"entity_type\": "
        "\"data_object\", \"operations\": \"[{\\\"operation\\\": \\\"add\\\", "
        "\\\"attribute\\\": \\\"attr\\\", \\\"value\\\": \\\"val\\\"}]\"}";
    
    int status;
    rodsEnv env;
    rErrMsg_t errmsg;
    rcComm_t *conn = NULL;

    status = getRodsEnv(&env);
    if (status != 0) {
        fprintf(stderr, "Failed to load iRODS environment\n");
        return 1;
    }

    conn = rcConnect(env.rodsHost, env.rodsPort, env.rodsUserName,
                     env.rodsZone, NO_RECONN, &errmsg);
    if (!conn) {
        fprintf(stderr, "Failed to connect\n");
        return 1;
    }
    char *results;
    
    status = rc_atomic_apply_metadata_operations(
                               conn, json_str, &results);
    rcDisconnect(conn);

    return status;
}
```

### Expected behavior
Metadata ("attr"="val") is placed on the object

### Observed behavior (including steps to reproduce, if applicable)
Locally: 
```
ERROR: apiTableLookup of apiNumber 20002 failed
```
and on the server:
```
Nov 11 15:31:56 pid:668 remote addresses: 127.0.0.1 ERROR: [-]    /irods/server/core/src/rsApiHandler.cpp:541:int readAndProcClientMsg(rsComm_t *, int) :  status [SYS_HEADER_READ_LEN_ERR]  errno [] -- message [failed to call 'read header']
    [-]    /irods/lib/core/src/sockComm.cpp:197:irods::error readMsgHeader(irods::network_object_ptr, msgHeader_t *, struct timeval *) :  status [SYS_HEADER_READ_LEN_ERR]  errno [] -- message [failed to call 'read header']
        [-]    /irods/plugins/network/tcp/libtcp.cpp:194:irods::error tcp_read_msg_header(irods::plugin_context &, void *, struct timeval *) :  status [SYS_HEADER_READ_LEN_ERR]  errno [] -- message [only read [0] of [4]]
```

Is this the correct way to use the api?

**Comment by korydraughn** [comment]
you need to call `load_client_api_plugins()` before `rc_atomic_apply_metadata_operations()`.

that function needs to be called only once. here is one example: https://github.com/irods/irods_client_icommands/blob/0a129b7c6a0f6e8104effb792cff7902bad1c504/src/itouch.cpp#L68

**Comment by korydraughn** [comment]
also, the JSON string does not match the required format.

the "operations" JSON property should not be treated as a string. try the following:
```c++
const char *json_str = "{"
    "\"entity_name\": \"/testZone/home/irods/test_object\","
    "\"entity_type\": "\"data_object\","
    "\"operations\": [{"
        "\"operation\": \"add\","
        "\"attribute\": \"attr\","
        "\"value\": \"val\""
    "}]"
"}";
```

**Comment by mksanger** [comment]
Thanks, I've made both of those changes, but I unfortunately still get the same response `failed to call 'read_header'`

**Comment by korydraughn** [comment]
So, the lookup error has been resolved and all that remains is the read header issue, right?

**Comment by korydraughn** [comment]
How are you compiling the program? Can you provide the compiler flags, linker options, etc?

**Comment by kjsanger** [comment]
Hi, I have a similar setup to @mksanger and I'm seeing this message too when I try to run the test. I'm using the following code and makefile:

atomic-repro.c

```
#include "rodsClient.h"
#include <atomic_apply_metadata_operations.h>

int main() {
const char *json_str = "{"
    "\"entity_name\": \"/testZone/home/irods/README\","
    "\"entity_type\": \"data_object\","
    "\"operations\": [{"
                       "\"operation\": \"add\","
                       "\"attribute\": \"attr\","
                       "\"value\": \"val\""
                      "}]"
    "}";
    
    int status;
    rodsEnv env;
    rErrMsg_t errmsg;
    rcComm_t *conn = NULL;

    status = getRodsEnv(&env);
    if (status != 0) {
        fprintf(stderr, "Failed to load iRODS environment\n");
        return 1;
    }

    conn = rcConnect(env.rodsHost, env.rodsPort, env.rodsUserName,
                     env.rodsZone, NO_RECONN, &errmsg);
    if (!conn) {
        fprintf(stderr, "Failed to connect\n");
        return 1;
    }

    load_client_api_plugins();

    char *result;    
    status = rc_atomic_apply_metadata_operations(conn, json_str, &result);
    if (status < 0) {
        printf("irods error: %d\n", status);
        return 1;
    }

    printf("Result: %s\n", result);
    rcDisconnect(conn);

    return status;
}
```

```
CC = gcc 
CFLAGS = -g -std=c99 -Wall -Werror 
LIBS = -lirods_client -lirods_common -lstdc++ -lssl -lcrypto -lpthread -ldl -lm -lrt

PREFIX = /opt/irods
CPPFLAGS = -I$(PREFIX)/include -I$(PREFIX)/include/irods 
LDFLAGS = -L$(PREFIX)/lib -L$(PREFIX)/lib/irods/externals

all: atomic-repro

atomic-repro: atomic-repro.c
	$(CC) $($CFLAGS) $(CPPFLAGS) -o atomic-repro atomic-repro.c $(LDFLAGS) $(LIBS)

test: atomic-repro
	./atomic-repro

clean:
	rm atomic-repro

.PHONY: all test clean
```

I'm building against 4.2.10 (using gcc) and running against a 4.2.10 server. On the server side I see in the logs:

```
Nov 11 15:31:56 pid:668 DEBUG: load_plugin - calling plugin_factory() in [/usr/lib/irods/plugins/api/libirods_atomic_apply_metadata_operations_server.so]
Nov 11 15:31:56 pid:668 DEBUG: Added API number [20002] to the client API whitelist.
Nov 11 15:31:56 pid:668 DEBUG: load_plugin - loaded [irods_atomic_apply_metadata_operations_server]
Nov 11 15:31:56 pid:668 DEBUG: init_api_table :: adding 20002 - [api_atomic_apply_metadata_operations] - [irods_atomic_apply_metadata_operations_server]
```

which looks good. On the call I get:

```
Nov 11 15:31:56 pid:668 remote addresses: 127.0.0.1 ERROR: [-]    /irods/server/core/src/rsApiHandler.cpp:541:int readAndProcClientMsg(rsComm_t *, int) :  status [SYS_HEADER_READ_LEN_ERR]  errno [] -- message [failed to call 'read header']
    [-]    /irods/lib/core/src/sockComm.cpp:197:irods::error readMsgHeader(irods::network_object_ptr, msgHeader_t *, struct timeval *) :  status [SYS_HEADER_READ_LEN_ERR]  errno [] -- message [failed to call 'read header']
        [-]    /irods/plugins/network/tcp/libtcp.cpp:194:irods::error tcp_read_msg_header(irods::plugin_context &, void *, struct timeval *) :  status [SYS_HEADER_READ_LEN_ERR]  errno [] -- message [only read [0] of [4]]
```

So I think the same is happening for me. I can see there are unit tests for this in the iRODS test suite, but
I wasn't able to deduce from the what it is we're missing.

Thanks.

**Comment by korydraughn** [comment]
the code you posted compiled and ran successfully for me on ubuntu 16. i think your code is fine.

the issue must have something to do with the environment or the server.

i'd recommend doing the following:
- try applying that AVU using the PRC (the PRC supports the atomic metadata endpoint too).
- verify that the shared libraries match the iRODS server version.
- verify that the client's irods_environment file is what you expect.
- maybe try running the unit test to see if the behavior is different.

also, are you using packages provided by packages.irods.org or are you compiling iRODS from source?

**Comment by d-w-moore** [comment]
Is it just possible that this needs a clientLogin call just after the successful rcConnect ? Most icommands do something of this kind:
```
    status = clientLogin( conn );
    if ( status != 0 ) {
        rcDisconnect( conn );
        exit( 7 );
    }
```

**Comment by korydraughn** [comment]
The account I used was likely a rodsadmin account so that might explain my success. I'll give Dan's suggestion a try and post what happened.

**Comment by kjsanger** [comment]
> Is it just possible that this needs a clientLogin call just after the successful rcConnect ? Most icommands do something of this kind:
> 
> ```
>     status = clientLogin( conn );
>     if ( status != 0 ) {
>         rcDisconnect( conn );
>         exit( 7 );
>     }
> ```

Thanks for spotting my error. The login call was present the original code where we observed this,
but I accidentally omitted it when making the reduced test case. However, restoring it still yields the
same error (I've amended the code above to include it - that matches what I'm testing now).

**Comment by kjsanger** [comment]
I'm working my way through your recommndations listed above ... I'll report back soon. Thanks for your help.

**Comment by kjsanger** [comment]
I tried builiding the test program against the official 4.2.10 deb packages on Ubuntu bionic and found that it succeeds when run. It seems that the client library that we built ourselves has a problem.

On the client side, is this capability loaded from a plugin, as it is on the server? I wonder if the way we have relocated the libraries may have affected this (if that's what happens) e.g. our client is looking for plugins in the wrong place? No evidence - just an idea.

**Comment by kjsanger** [comment]
I've confirmed by strace on the irods server with stock iRODS dev/runtime packages (where things work) that I can see the plugins loaded, while using the client libraries that we built, the plugins (except essentail network etc for connection) do not get loaded.

**Comment by kjsanger** [comment]
I have tested another scenario. Still building the above test program against stock 4.2.10 irods-dev, irods-runtime and irods-externals deb packages, but this time executing the test on a host that does _not_ have the irods server installed.

```
 ./atomic-repro 
remote addresses: 127.0.0.1 ERROR: procApiRequest: apiTableLookup of apiNumber 20002 failed
```

strace confirms the API plugins are not loaded in this case, much like the case where we built the client libraries. This is the main difference I'm observing. Is the client expected to load the API plugins?

**Comment by korydraughn** [comment]
Yes. The client is expected to load them. `load_client_api_plugins()` loads them for you.

**Comment by trel** [comment]
This highlights we may need a new/additional section in the upcoming 4.3 developers section of our docs...

https://github.com/irods/irods_docs/tree/master/docs/developers

Suggestions / PRs welcome.

**Comment by kjsanger** [comment]
> Yes. The client is expected to load them. `load_client_api_plugins()` loads them for you.

The problem seem to be that the API plugins are not packaged for the client - only the server has them:

```
root@tao:/opt/docker/irods# dpkg -L irods-runtime | grep api
root@tao:/opt/docker/irods# dpkg -L irods-server | grep api
/usr/lib/irods/plugins/api
/usr/lib/irods/plugins/api/libhelloworld_client.so
/usr/lib/irods/plugins/api/libhelloworld_server.so
/usr/lib/irods/plugins/api/libirods_atomic_apply_acl_operations_client.so
/usr/lib/irods/plugins/api/libirods_atomic_apply_acl_operations_server.so
/usr/lib/irods/plugins/api/libirods_atomic_apply_metadata_operations_client.so
/usr/lib/irods/plugins/api/libirods_atomic_apply_metadata_operations_server.so
/usr/lib/irods/plugins/api/libirods_data_object_finalize_client.so
/usr/lib/irods/plugins/api/libirods_data_object_finalize_server.so
/usr/lib/irods/plugins/api/libirods_data_object_modify_info_client.so
/usr/lib/irods/plugins/api/libirods_data_object_modify_info_server.so
/usr/lib/irods/plugins/api/libirods_get_file_descriptor_info_client.so
/usr/lib/irods/plugins/api/libirods_get_file_descriptor_info_server.so
/usr/lib/irods/plugins/api/libirods_replica_close_client.so
/usr/lib/irods/plugins/api/libirods_replica_close_server.so
/usr/lib/irods/plugins/api/libirods_replica_open_client.so
/usr/lib/irods/plugins/api/libirods_replica_open_server.so
/usr/lib/irods/plugins/api/libirods_touch_client.so
/usr/lib/irods/plugins/api/libirods_touch_server.so
```

**Comment by SwooshyCueb** [comment]
We have moved the api plugins to the runtime package for 4.2.11: #5836

Can you confirm that the issue goes away when the irods-server package is installed? If so, I think we can close this issue.

**Comment by kjsanger** [comment]
Yes, when using the deb packages, adding the server allows the API plugins to be found and loaded.

I've clearly done something wrong packaging our Conda build because while our runtime package includes
the API plugins and the above test program resolves the installed plugin directory, I still see a failure to load them, which falls into the category of "my problem, not yours".

However, it would be very helpful to know if there's any reason why that approach shouldn't work? Should it be sufficient for the client to resolve the root plugin directory (i.e. the one directly containing the "api" subdirectory) and for the plugins to be in the "api" subdirectory, for loading to occur?

Thanks.

**Comment by SwooshyCueb** [comment]
We did recently change how iRODS locates installation directories, but I can't remember offhand if that made it into 4.2.10. I'll have a look shortly and see if I can't offer some guidance.

**Comment by SwooshyCueb** [comment]
Starting in 4.2.9, the way default paths are handled was changed:
https://github.com/irods/irods/commit/b06b6c0b1fa44c1ad4925f6d011ce3c19b3afbde

Where previously we made assumptions about installation paths, we now use CMake variables to try and derive where things might live. We do so in such a manner as to try and tolerate re-prefixing (i.e. you should be able to move the entire irods installation while preserving the basic directory structure defined at configure time without causing any problems).
As this derivation takes place in a runtime library, we assume our starting point is in `CMAKE_INSTALL_LIBDIR`, so we can match up the absolute path of the runtime library's installation directory with the value of  `CMAKE_INSTALL_LIBDIR`, which is typically a relative path. We can then calculate the relative path from `CMAKE_INSTALL_LIBDIR` to `IRODS_PLUGINS_DIRECTORY`, to get the absolute path of `IRODS_PLUGINS_DIRECTORY`. However, if `IRODS_PLUGINS_DIRECTORY` is *already* an absolute path, we just use it as-is. Defining any of these CMake variables as absolute paths will usually break the re-prefixing tolerance, but if you know that the files will always live in the same place, it shouldn't be an issue.

I hope that makes sense.

**Comment by kjsanger** [comment]
Thanks, that's very helpful. After making a debug build, I can see why the plugins are not being loaded in our Conda package - it's a case of the plugins path being set up differently under Conda. The good news is that it looks like it is a straightforward fix to get our clients doing atomic updates.

**Comment by trel** [comment]
@kjsanger let us know when you get things loading and we'll close this as a success.

**Comment by korydraughn** [comment]
@kjsanger Any updates on this?

**Comment by kjsanger** [comment]
Yes, we patch iRODS to get this to work. The exact patch has changed with versions, but for 4.2.11 and 4.3.0 is identical:

https://github.com/wtsi-npg/npg_conda/blob/devel/recipes/irods/4.2.11/irods-lib-core-irods-load-plugin-path-truncation-bug.patch

We no longer get a problem with path construction now that CMake knows where things are, but we do see a rogue NUL in the path when converted to a C string, which the patch addresses.

**Comment by korydraughn** [comment]
Great. I'll go ahead and close this.

If you experience other problems. Please open a new issue.
